sysl2/Grammar1_test.go

```diff
@@ -0,0 +1,253 @@
+package main
+
+import (
+    "testing"
+
+    sysl "anz-bank/sysl/sysl2/proto"
+)
+
+func testParser(g *sysl.Grammar, numTerms int, tokens []int, text string, expected bool, t *testing.T) {
+    p := makeParser(g, text)
+
+    if len(p.terminals) != numTerms {
+        t.Error("got incorrect number of terms", p.terminals)
+    }
+
+    for i, expected := range tokens {
+        tok := p.l.nextToken()
+        if tok != expected {
+            t.Errorf("got the wrong token at %d expected: %d, got %d", i, expected, tok)
+        }
+    }
+
+    result := p.parse(tokens[:len(tokens)-1])
+
+    if result != expected {
+        t.Error("failed to parse " + text)
+    }
+}
+
+// S –> bab | bA
+// A –> d | cA
+func TestGrammar(t *testing.T) {
+    tokens := []int{2, 3, 2, -1}
+    text := "bab"
+
+    testParser(makeGrammar1(), 4, tokens, text, true, t)
+}
+
+func TestGrammar1(t *testing.T) {
+    tokens := []int{2, 1, 1, 1, 0, -1}
+    text := "bcccd"
+
+    testParser(makeGrammar1(), 4, tokens, text, true, t)
+}
+
+func TestGrammarIncompleteInput(t *testing.T) {
+    tokens := []int{2, 3, -1}
+    text := "ba"
+
+    testParser(makeGrammar1(), 4, tokens, text, false, t)
```

Comment:
I'm a bit confused about what the `4` is. You call it `numTerms ` above, but it doesn't seem to change depending on the input text, so it seems to be something inherent in the grammar definition itself? If that's true that I think it would have been better to write a separate test that only checks for qualities of the grammar without input text.